tests/integration/ha_tests/helpers.py

```diff
@@ -0,0 +1,337 @@
+        return pod.status.podIP
+
+    app = await app_name(ops_test)
+    member_ips = {}
+    for unit in ops_test.model.applications[app].units:
+        unit_address = await get_unit_address(ops_test, unit.name)
+        cluster_info = requests.get(f"http://{unit_address}:8008/cluster")
+        if len(member_ips) > 0:
+            # If the list of members IPs was already fetched, also compare the
+            # list provided by other members.
+            assert member_ips == {
+                get_host_ip(member["host"]) for member in cluster_info.json()["members"]
+            }, "members report different lists of cluster members."
+        else:
+            member_ips = {get_host_ip(member["host"]) for member in cluster_info.json()["members"]}
+    return member_ips
+
+
+async def get_master_start_timeout(ops_test: OpsTest) -> Optional[int]:
+    """Get the master start timeout configuration.
+
+    Args:
+        ops_test: ops_test instance.
+
+    Returns:
+        master start timeout in seconds or None if it's using the default value.
+    """
+    for attempt in Retrying(stop=stop_after_delay(30 * 2), wait=wait_fixed(3)):
+        with attempt:
+            app = await app_name(ops_test)
+            primary_name = await get_primary(ops_test, app)
+            unit_ip = await get_unit_address(ops_test, primary_name)
+            configuration_info = requests.get(f"http://{unit_ip}:8008/config")
+            master_start_timeout = configuration_info.json().get("master_start_timeout")
+            return int(master_start_timeout) if master_start_timeout is not None else None
+
+
+async def get_password(ops_test: OpsTest, app) -> str:
+    """Use the charm action to retrieve the password from provided application.
+
+    Returns:
+        string with the password stored on the peer relation databag.
+    """
+    # Can retrieve from any unit running unit, so we pick the first.
+    for attempt in Retrying(stop=stop_after_attempt(len(ops_test.model.applications[app].units))):
+        with attempt:
+            unit_name = (
+                ops_test.model.applications[app].units[attempt.retry_state.attempt_number - 1].name
+            )
+            action = await ops_test.model.units.get(unit_name).run_action("get-password")
+            action = await asyncio.wait_for(action.wait(), 10)
+            return action.results["operator-password"]
+
+
+async def get_primary(ops_test: OpsTest, app) -> str:
+    """Use the charm action to retrieve the primary from provided application.
+
+    Returns:
+        string with the password stored on the peer relation databag.
+    """
+    # Can retrieve from any unit running unit, so we pick the first.
+    for attempt in Retrying(stop=stop_after_attempt(len(ops_test.model.applications[app].units))):
+        with attempt:
+            unit_name = (
+                ops_test.model.applications[app].units[attempt.retry_state.attempt_number - 1].name
+            )
```

Comment:
If the PostGres service has been paused on a unit and then you ask that unit for the password will this still work? 

If not might want to change which unit you're grabbing to be any unit except the one that was paused

Reply:
This mechanism was already handling that situation (getting the info from a different unit when a unit is paused).

However, I updated the code on https://github.com/canonical/postgresql-k8s-operator/pull/59/commits/ac65418440fda0771a07d2563a5b6bc6fc329f87 (along with some other changes) to have a better implementation like we have in the VM charm and other charms.